woocommerce/woocommerce-android
# Feature/image upload errors

This PR includes all the changes from #4411 , #4432 , #4435. The changes in this PR have already been reviewed individually and this is just to merge the feature branch into `develop`.

### Changes
- This PR modifies the text on the SnackBar that is displayed when an image upload process fails. Clicking on the `Details` button on the error snackbar displayed when image upload fails, redirects to a new screen. This screen displays the list of images that failed to upload, along with the error message. 
- I introduced a new class called `MediaFileUploadHandler` that keeps track of the number of images that failed to upload and the reason for that error. This makes it easier to unit test this functionality. I do think we should refactor `ProductImagesService` and move all the logic in that service class to this class at some point so that it is easier to test but that is out of scope for this project atleast. I'll open a separate issue for that. 
- We also allow the merchant to redirect to the Product detail screen when the back button is clicked, when images are being uploaded. If an image upload fails in the Product detail or Variation detail screen, a similar Snackbar would be displayed with a `Details` button. Clicking on the `Details` would redirect to a new screen that displays the list of images that failed to upload.

### Screenshots
##### Product Detail redirection | Variation Detail redirection
<img src="https://user-images.githubusercontent.com/22608780/125621739-6626e6ce-49ad-4d4c-a0bc-7812facada02.gif" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/125621939-1f1f245c-8436-4671-87a5-05a9a8549528.gif" width="300"/>

#### Error Details screen
<img src="https://user-images.githubusercontent.com/22608780/125584213-b6814c0c-2769-4d52-8585-d37d8f38ffe7.gif" width="300"/>

### Testing
- Try to upload an image with an extension, other than `png`, `jpeg`, `jpg` or `gif`. There is local validation in FluxC so any images without these extensions will result in an error.
- While the image is being uploaded, click on the back button. You should be redirected to the Product Details screen.
- Verify that the error snackbar is displayed with a DETAILS button. Click on the `Details` button and note that you are redirected to a new screen that displays a list of errors.
- Follow the above steps for a variation of a product.

Update release notes:
